source/GameInterface/Services/Clans/Patches/ClanChangeKingdomPatch.cs

```diff
@@ -16,39 +16,33 @@ namespace GameInterface.Services.Clans.Patches
     [HarmonyPatch(typeof(ChangeKingdomAction), "ApplyInternal")]
     public class ClanChangeKingdomPatch
     {
-        private static readonly AllowedInstance<Clan> AllowedInstance = new AllowedInstance<Clan>();
-
         private static readonly Action<Clan, Kingdom, ChangeKingdomAction.ChangeKingdomActionDetail, int, bool, bool> ApplyInternal =
             typeof(ChangeKingdomAction)
             .GetMethod("ApplyInternal", BindingFlags.NonPublic | BindingFlags.Static)
             .BuildDelegate<Action<Clan, Kingdom, ChangeKingdomAction.ChangeKingdomActionDetail, int, bool, bool>>();
 
         static bool Prefix(Clan clan, Kingdom newKingdom, ChangeKingdomAction.ChangeKingdomActionDetail detail, int awardMultiplier = 0, bool byRebellion = false, bool showNotification = true)
         {
-            if (AllowedInstance.IsAllowed(clan)) return true;
+            if (AllowedThread.IsThisThreadAllowed()) return true;
 
             if (PolicyProvider.AllowOriginalCalls) return true;
 
             if (ModInformation.IsClient && clan != Clan.PlayerClan) return false;
 
-            CallStackValidator.Validate(clan, AllowedInstance);
-
             MessageBroker.Instance.Publish(clan, new ClanKingdomChanged(clan.StringId, newKingdom?.StringId, (int)detail, awardMultiplier, byRebellion, showNotification));
 
             return false;
         }
 
         public static void RunOriginalChangeClanKingdom(Clan clan, Kingdom newKingdom, ChangeKingdomAction.ChangeKingdomActionDetail detail, int awardMultiplier = 0, bool byRebellion = false, bool showNotification = true)
         {
-            using (AllowedInstance)
+            GameLoopRunner.RunOnMainThread(() =>
             {
-                AllowedInstance.Instance = clan;
-
-                GameLoopRunner.RunOnMainThread(() =>
+                using(new AllowedThread())
                 {
                     ApplyInternal.Invoke(clan, newKingdom, detail, awardMultiplier, byRebellion, showNotification);
-                }, true);
-            }
+                }
+            }, true);
```

Comment:
Can you update this to non-blocking in your branch so we dont forget
```suggestion
            });
```